documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/RoutableFactories80.java

```diff
@@ -0,0 +1,931 @@
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * Implementation of MessageBus message request/response serialization built around Protocol Buffers.
+ */
+abstract class RoutableFactories80 {
+
+    private static class ProtobufCodec<DocApiT extends Routable, ProtoT extends AbstractMessage> implements RoutableFactory {
+
+        private final Class<DocApiT>                          apiClass;
+        private final Function<DocApiT, ProtoT>               encoderFn;
+        private final Function<DocumentDeserializer, DocApiT> decoderFn;
+
+        ProtobufCodec(Class<DocApiT> apiClass,
+                      Function<DocApiT, ProtoT> encoderFn,
+                      Function<DocumentDeserializer, DocApiT> decoderFn) {
+            this.apiClass = apiClass;
+            this.encoderFn = encoderFn;
+            this.decoderFn = decoderFn;
+        }
+
+        @Override
+        public boolean encode(Routable obj, DocumentSerializer out) {
+            try {
+                var protoMsg = encoderFn.apply(apiClass.cast(obj));
+                var protoStream = CodedOutputStream.newInstance(out.getBuf().getByteBuffer()); // Not AutoCloseable...
+                try {
+                    protoMsg.writeTo(protoStream);
+                } finally {
+                    protoStream.flush();
+                }
+            } catch (IOException | UnsupportedOperationException e) {
+                return false;
+            }
+            return true;
+        }
+
+        @Override
+        public Routable decode(DocumentDeserializer in) {
+            return decoderFn.apply(in);
+        }
+    }
+
+    private static class ProtobufCodecBuilder<DocApiT extends Routable, ProtoT extends AbstractMessage> {
+
+        private final Class<DocApiT>                    apiClass;
+        private final Class<ProtoT>                     protoClass;
+        private Function<DocApiT, ProtoT>               encoderFn;
+        private Function<DocumentDeserializer, DocApiT> decoderFn;
+
+        ProtobufCodecBuilder(Class<DocApiT> apiClass, Class<ProtoT> protoClass) {
+            this.apiClass = apiClass;
+            this.protoClass = protoClass;
+        }
+
+        static <DocApiT extends Routable, ProtoT extends AbstractMessage> ProtobufCodecBuilder<DocApiT, ProtoT>
+        of(Class<DocApiT> apiClass, Class<ProtoT> protoClass) {
+            return new ProtobufCodecBuilder<>(apiClass, protoClass);
+        }
+
+        ProtobufCodecBuilder<DocApiT, ProtoT> encoder(Function<DocApiT, ProtoT> fn) {
+            if (encoderFn != null) {
+                throw new IllegalArgumentException("Encoder already set");
+            }
+            encoderFn = fn;
+            return this;
+        }
+
+        ProtobufCodecBuilder<DocApiT, ProtoT> decoder(ProtoT messagePrototype, Function<ProtoT, DocApiT> fn) {
```

Comment:
Did you consider just passing the `parser()` here, instead of the `getDefaultInstance()`? It's possible this is the idiomatic way with protobuf, but that's not obvious to me. That also lets you get rid of the cast on L108. 
Similarly for `decoderWithRepo(...)`.

Reply:
Have changed the code to use the underlying parser directly